Simplify attaching to raw tracepoints

buglloc opened the pull request:
First of al,l thanks for the great project! :+1: 

I use `github.com/cilium/ebpf` to monitor various system events through raw tracepoints. This PR helps to use it (not needed to copy-paste BPF syscall wrapper and other).

lmb commented:
Glad it's useful for you :)

Take a look at #83 where we discussed a similar proposal. The gist: there are many attachment points of BPF, and supporting them all sounds like a burden.

OTOH I see that this is only a small amount of code, and it would make the library more useful out of the box. Maybe we can put it in a subpackage? That way it could still access the internal stuff. @tklauser @florianl @joestringer thoughts?

This would raise another question: do we add a separate subpackage for #83 ? Or should that be unified with raw tracepoints (what is the difference)? Plus, can we move Program.Attach, Program.Detach to a subpackage as well? It seems way less general than I remembered, especially now that bpf_link is supposed to come and be the new thing.

florianl commented:
I think we should not mix it with https://github.com/cilium/ebpf/issues/83. As far as I know tracepoints are different to raw tracepoints from eBPF point of view. To attach eBPF programs to regular tracepoints you have to use `ioctl` while there is a own struct [1] for raw tracepoints in the `union bpf_attr`. And I agree that the handling of `ioctl` is out of scope of this package.
But attaching a eBPF program to a raw tracepoint should be part of this package or a subpackage, I think.
With bpf link on the horizon, I think, it should coexit with existing attach/detach methods.

[1] https://github.com/torvalds/linux/blob/444565650a5fe9c63ddf153e6198e31705dedeb2/include/uapi/linux/bpf.h#L546-L549

lmb commented:
I see. Hooray for consistent and easy to understand naming...

@buglloc do you feel like moving the function into a subpackage? `raw_tracepoints` maybe? With some documentation that the package may vanish at any point in time :P

buglloc commented:
> @buglloc do you feel like moving the function into a subpackage? raw_tracepoints maybe? With some documentation that the package may vanish at any point in time :P

Yessir!

lmb commented:
I'm currently looking into #102, and I think that will in some way involve adding support for raw tracepoints (not entirely sure yet). Just thought I'd give you a heads up so that you don't end up wasting your time.

florianl commented on raw_tracepoints/attach.go:
```diff
@@ -0,0 +1,23 @@
+// Attach a Program fd to a raw tracepoint.
+//
+// Requires at least Linux 4.17.
+func AttachFD(fd int, tpName string) error {
```
Can you update the comment and use the commonly used style?
```
// AttachFD ...
```

florianl commented on raw_tracepoints/attach_test.go:
```diff
@@ -0,0 +1,31 @@
+)
+
+func TestProgramAttachRawTracepoint(t *testing.T) {
+	testutils.SkipOnOldKernel(t, "4.17", "BPF_RAW_TRACEPOINT Api")
```
Can we capitalize `API`?

florianl commented on raw_tracepoints/syscalls.go:
```diff
@@ -0,0 +1,21 @@
+)
+
+const (
+	rawTracepointOpen = 17
```
To avoid import loops, we can not use `_RawTracepointOpen` from [1]. But keeping the same name schema would be great.

[1] https://github.com/cilium/ebpf/blob/master/types.go#L112

buglloc commented:
Sorry for the long delay, have a lot of work :(

Fixed previous issues and moved raw tracepoint to the `link` subpackage (as far as i'm concerned, it's more appropriate for now)

florianl commented on link/syscalls.go:
```diff
@@ -134,3 +134,17 @@ func bpfIterCreate(attr *bpfIterCreateAttr) (*internal.FD, error) {
+type bpfRawTracepointOpenAttr struct {
+	name internal.Pointer
+	fd   uint32
+	pad  uint32
```
`pad` is not part of the `BPF_RAW_TRACEPOINT` `bpf_attr`. Can we use `_` instead?

buglloc replied:
Oh, sure. my bad :)

buglloc commented:
Thanks! :)

tklauser commented:
Thanks for your contribution @buglloc :rocket:

lmb commented on link/raw_tracepoint.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+func (rt *progAttachRawTracepoint) Update(_ *ebpf.Program) error {
+	return fmt.Errorf("can't update raw_tracepoint: %w", ErrNotSupported)
```
Is there no way to support `Update`? I would like progAttachRawTracepoint to pass `testLink()` which currently fails due to this.

buglloc replied:
Honestly, I wasn't sure if it was useful. Implemented it in #122 (and uses `testLink` in tests)
